[torch.package][easy] Add another opcode for matching pickle protocol 4+ correctly

kunalb opened the pull request:
Summary: IL generates massive function names: which meant that the pickle opcode used is BINUNICODE instead of the short version -- and then it would silently get skipped while pickling with protocol 4.

Differential Revision: D44815351

ezyang commented:
Can you add a test?

ezyang commented:
Without understanding the underlying logic here, this change is confusing to me. SHORT_BINUNICODE is Protocol 4 only. The test for BINUNICODE8 seems parallel, as it is also Protocol 4 only. With the change you have made here, you now have a mix of Protocol 4 and Protocol 1. Smells fishy.

kunalb commented:
Ah, you're right; for some reason I didn't see binunicode8 but just saw binunicode when I looked at pickletools. The generated ops I see for a very long function name return binunicode instead of binunicode8, will add a test.

kunalb commented:
![image](https://user-images.githubusercontent.com/377991/230751240-c76fe7bd-8899-49ec-93dc-6efbad811d00.png)

Just to make sure I wasn't hallucinating: opcodes for pickling a function with a long name. I'll take a look at pickle's implementation to see how/when binunicode vs binunicode8 are chosen.

kunalb commented:
So I guess pickle goes from 
SHORT_BIN_UNICODE (<256 characters) -> 
BINUNICODE (<4294967295 characters or < protocol 4) ->
BINUNICODE8 (> protocol 4 and > 4....5 characters)

https://github.com/python/cpython/blob/3.7/Lib/pickle.py#L739
```
 def save_str(self, obj):
        if self.bin:
            encoded = obj.encode('utf-8', 'surrogatepass')
            n = len(encoded)
            if n <= 0xff and self.proto >= 4:
                self.write(SHORT_BINUNICODE + pack("<B", n) + encoded)
            elif n > 0xffffffff and self.proto >= 4:
                self._write_large_bytes(BINUNICODE8 + pack("<Q", n), encoded)
            elif n >= self.framer._FRAME_SIZE_TARGET:
                self._write_large_bytes(BINUNICODE + pack("<I", n), encoded)
            else:
                self.write(BINUNICODE + pack("<I", n) + encoded)
        else:
            obj = obj.replace("\\", "\\u005c")
            obj = obj.replace("\0", "\\u0000")
            obj = obj.replace("\n", "\\u000a")
            obj = obj.replace("\r", "\\u000d")
            obj = obj.replace("\x1a", "\\u001a")  # EOF on DOS
            self.write(UNICODE + obj.encode('raw-unicode-escape') +
                       b'\n')
        self.memoize(obj)
```

fluckydog232 commented:
checked there is a linter signal failure, but given the nature of  tests added(abnormal size string), it make sense to override the check.

ezyang commented on torch/package/package_exporter.py:
```diff
@@ -670,6 +670,7 @@ def _check_mocked_error(module: Optional[str], field: Optional[str]):
                 if pickle_protocol == 4:
                     if (
                         opcode.name == "SHORT_BINUNICODE"
+                        or opcode.name == "BINUNICODE"
```
I am still perplexed. Don't we have to do the BINUNICODE special handling for older pickle protocol too?

kunalb replied:
Ah, no: this was actually a change in between 3 & 4
https://peps.python.org/pep-3154/#serializing-more-lookupable-objects

STACK_GLOBAL was added as part of 4 that takes the topmost stack items to figure out the values and looks up the strings  as module_name and qual_name (so this code basically keeps a running buffer of the last 2 strings seen just in case it runs into a STACK_GLOBAL)

In pickle 3 there was just GLOBAL that accepts the module & field as an argument -- so no separate string handling/no custom binunicodes, handled in the if/else around line 700.
https://github.com/python/cpython/blob/main/Lib/pickletools.py#L1928

ezyang replied:
ok!

kunalb-meta committed:
[torch.package][easy] Add opcode for strings making packaging behave incorrectly for protocol 4. (#98674)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/98674

IL generates massive function names: which meant that the pickle opcode used is BINUNICODE instead of the short version -- and then it would silently get skipped while pickling with protocol 4.

Reviewed By: fluckydog232, atuljangra

Differential Revision: D44815351

fbshipit-source-id: 423027765893737f3fc2a21ba6de249060b7cddb
